RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -11,261 +8,421 @@ namespace {
-    _condnames_mean_25ns.push_back("ecalPFClusterCorV2_EE_pfSize1_mean_25ns");
-    _condnames_mean_25ns.push_back("ecalPFClusterCorV2_EE_pfSize2_mean_25ns");
-    _condnames_mean_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin1_mean_25ns");
-    _condnames_mean_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin2_mean_25ns");
-    _condnames_mean_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin3_mean_25ns");
+    for (unsigned int icor=0; icor<ncor; ++icor) {
+      es.get<GBRDWrapperRcd>().get(condnames_mean[icor],forestH_mean[icor]);
+      es.get<GBRDWrapperRcd>().get(condnames_sigma[icor],forestH_sigma[icor]);
+    }
     
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EB_pfSize1_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EB_pfSize2_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EB_pfSize3_ptbin1_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EB_pfSize3_ptbin2_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EB_pfSize3_ptbin3_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EE_pfSize1_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EE_pfSize2_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin1_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin2_sigma_25ns");
-    _condnames_sigma_25ns.push_back("ecalPFClusterCorV2_EE_pfSize3_ptbin3_sigma_25ns");
-  }
-  
-
-  
-}
-
-
-void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const edm::EventSetup &es, const reco::PFCluster::EEtoPSAssociation &assoc, reco::PFClusterCollection& cs) {
-
-  //legacy corrections
-  if (!_applyMVACorrections) {
-    for (unsigned int idx = 0; idx<cs.size(); ++idx) {
+    std::array<float,5> eval;            
+    for (unsigned int idx = 0; idx<cs.size(); ++idx) {                        
       reco::PFCluster &cluster = cs[idx];
       bool iseb = cluster.layer() == PFLayer::ECAL_BARREL;
+      float ePS1=0., ePS2=0.;
+      if(!iseb)
+	getAssociatedPSEnergy(idx, assoc, ePS1, ePS2);
+
+      double e = cluster.energy();
+      double pt = cluster.pt();
+      double evale = e;
+      if (maxPtForMVAEvaluation_>0. && pt>maxPtForMVAEvaluation_) {
+	evale *= maxPtForMVAEvaluation_/pt; 
+      }
+      double invE = (e == 0.) ? 0. : 1./e; //guard against dividing by 0.      
+      int size = lazyTool.n5x5(cluster);      
+      
+      int ietaix=0;
+      int iphiiy=0;
+      if (iseb) {
+	EBDetId ebseed(cluster.seed());
+	ietaix = ebseed.ieta();
+	iphiiy = ebseed.iphi();
+      } else {
+	EEDetId eeseed(cluster.seed());
+	ietaix = eeseed.ix();
+	iphiiy = eeseed.iy();      
+      }
+
+      //find index of corrections (0-4 for EB, 5-9 for EE, depending on cluster size and raw pt)
+      int coridx = std::min(size,3)-1;
+      if (coridx==2) {
+	if (pt>4.5) {
```

Comment:
the magic numbers to select pt are worrisome - are these somehow promised never to change?